### PR TITLE
Refactor CI/CD to decrease iteration time

### PR DESCRIPTION
**Pull request by Alan-Jowett**
Signed-off-by: Alan Jowett <alanjo@microsoft.com>

## Description

Create a single unified CI/CD workflow for eBPF for Windows.

## Testing

Verify code coverage numbers are reported.
Verify tests are executed.

## Documentation

No.

Resolves: #898
Resolves: #899 
Resolves: #886 
Resolves: #856

### CODE REVIEW

File: .github/workflows/cicd.yml [hunk 1]
```diff
@@ -0,0 +1,110 @@
+# Copyright (c) Microsoft Corporation
+# SPDX-License-Identifier: MIT
+
+# This is the top-level workflow that runs on each pull request and push to main.
```
Comment (dthaler) [inline .github/workflows/cicd.yml:4]: Why do we need to run it on a push to main?  (Add answer in the file comment.)  It should have already run on the pull request being merged to main, no?

File: .github/workflows/cicd.yml [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+      pre_test: call "C:\Program Files (x86)\Microsoft Visual Studio\2019\Enterprise\Common7\Tools\VsDevCmd.bat"
+      test_command: bpf2c_tests.exe -s
+      name: bpf2c
+      build_job: regular / build
+      build_artifact: Build-x64
+      environment: windows-2019
+      code_coverage: true
+      gather_dumps: true
+
+  # Run the driver tests on self-hosted runners.
```
Comment (dthaler) [inline .github/workflows/cicd.yml:90]: nit: this one ends in a period, the others don't.

File: .github/workflows/reusable-test.yml [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+      build_job:
+        required: true
+        type: string
+      # The name of the build artifact to download.
+      build_artifact:
+        required: true
+        type: string
+      # The environment to run this test on.
+      environment:
+        description: 'Environment to run tests against'
```
Comment (dthaler) [inline .github/workflows/reusable-test.yml:28]:
```suggestion
        description: 'Environment to run tests on'
```
(wording consistency with line 26)

File: .github/workflows/reusable-test.yml [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+        type: string
+        required: true
+      # Set to true to gather code coverage when this test runs.
+      code_coverage:
+        required: false
+        type: boolean
+      #  Set to true to gather memory and upload dumps if a test process crashes.
```
Comment (dthaler) [inline .github/workflows/reusable-test.yml:35]:
```suggestion
      # Set to true to gather memory and upload dumps if a test process crashes.
```

File: .github/workflows/cicd.yml (hunk 1 above)
Comment (Alan-Jowett) [reply .github/workflows/cicd.yml:4]: The answer is on line 10. When we do a squash commit to main from a PR, it creates a new unique commit. Codecov.io requires a commit as a baseline to compare against, but in the case of a squash the PR isn't the same commit as the one made to main.

File: .github/workflows/cicd.yml [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+      environment: windows-2019
+      code_coverage: true
+      gather_dumps: true
+
+  # Run the API tests in GitHub.
+  ebpf_client:
+    uses: ./.github/workflows/reusable-test.yml
+    with:
+      name: ebpf_client
+      test_command: ebpf_client.exe -s
```
Comment (shankarseal) [inline .github/workflows/cicd.yml:58]: remove this. Redundant as it runs in selfhost runner. In github runner it wont exercise the kernel code path. The mock code path is executed by unit test anyways.

Comment (Alan-Jowett) [reply .github/workflows/cicd.yml:58]: Sure, happy to remove this after we add the code coverage driver test run.

File: .github/workflows/reusable-test.yml [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+        required: false
+        type: string
+
+permissions:
+  checks: read  # Required by fountainhead/action-wait-for-check to wait for another GitHub check to complete.
+  contents: read  # Required by actions/checkout to fetch code.
+
+jobs:
+  run_test:
+    timeout-minutes: 30
```
Comment (shankarseal) [inline .github/workflows/reusable-test.yml:52]: does this resolve #856  ?

Comment (Alan-Jowett) [reply .github/workflows/reusable-test.yml:52]: Yes, it should.

File: .github/workflows/reusable-test.yml [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+    - uses: actions/checkout@b0e28b5ac45a892f91e7d036f8200cf5ed489415
+      # Only check out source code if code coverage is being gathered.
+      if: inputs.code_coverage == true
+      with:
+        submodules: 'recursive'
+        ref: ${{ github.event.workflow_run.head_branch }}
+
+    - name: Set up OpenCppCoverage and add to PATH
+      id: set_up_opencppcoverage
+      if: inputs.code_coverage == true
```
Comment (shankarseal) [inline .github/workflows/reusable-test.yml:78]: code_coverage & not selfhost_runner?

File: tests/bpf2c_tests/raw_bpf.cpp [hunk 8]
```diff
@@ -39,7 +39,7 @@ run_test(const std::string& data_file)
         state_result,
```
Comment (shankarseal) [inline tests/bpf2c_tests/raw_bpf.cpp:39]: recommend a separate PR?

File: tests/sample/encap_reflect_packet.c [hunk 9]
```diff
@@ -3,7 +3,7 @@
 
 #include "xdp_common.h"
 
-int
+__attribute__((always_inline)) int
```
Comment (shankarseal) [inline tests/sample/encap_reflect_packet.c:6]: what is the reason for this change?

File: .github/workflows/cicd.yml [hunk 10]
```diff
@@ -0,0 +1,110 @@
...
+
+  # Run the driver tests on self-hosted runners.
+  driver:
+    uses: ./.github/workflows/reusable-test.yml
+    with:
+      test_command: powershell ".\setup_ebpf_cicd_tests.ps1; .\execute_ebpf_cicd_tests.ps1; .\cleanup_ebpf_cicd_tests.ps1"
+      name: driver
+      build_job: regular / build
+      build_artifact: Build-x64
+      environment: ebpf_cicd_tests
```
Comment (shankarseal) [inline .github/workflows/cicd.yml:98]: code_coverage and gather_dumps should be true for this as well. Note that the tools are already present in the VMs.

File: tests/sample/encap_reflect_packet.c (hunk 9 above)
Comment (Alan-Jowett) [reply tests/sample/encap_reflect_packet.c:6]: Switching to using LLVM 13.0.0 which is already present on the runner. LLVM 13.0.0.0 doesn't inline this without this attribute, causing a verifier failure.

File: .github/workflows/reusable-test.yml [hunk 11]
```diff
@@ -0,0 +1,168 @@
...
+        required: false
+        type: boolean
+      # Set to true to gather and upload memory dumps if a test process crashes.
+      gather_dumps:
+        required: false
+        type: boolean
+      pre_test:
+        required: false
+        type: string
+      post_test:
```
Comment (shankarseal) [inline .github/workflows/reusable-test.yml:42]: not used. Did you want it for some of the top level jobs?

File: .github/workflows/cicd.yml (hunk 10 above)
Comment (Alan-Jowett) [reply .github/workflows/cicd.yml:98]: Yes on the dumps. Will need to do more stuff to get the code coverage working as part of another PR.

File: .github/workflows/cicd.yml [hunk 12]
```diff
@@ -90,6 +90,8 @@ jobs:
   # Run the driver tests on self-hosted runners.
   driver:
     uses: ./.github/workflows/reusable-test.yml
+    # Only run this on repo's that have self-host runners.
```
Comment (dthaler) [inline .github/workflows/cicd.yml:93]:
```suggestion
    # Only run this on repos that have self-host runners.
```

File: tests/sample/sample.vcxproj [hunk 13]
```diff
@@ -336,17 +336,6 @@
       <TreatOutputAsContent Condition="'$(Configuration)|$(Platform)'=='Release|x64'">true</TreatOutputAsContent>
     </CustomBuild>
   </ItemGroup>
-  <ItemGroup>
-    <CustomBuild Include="printk_legacy.c">
```
Comment (dthaler) [inline tests/sample/sample.vcxproj:340]: why is this one removed?

File: tests/sample/sample.vcxproj.filters [hunk 14]
```diff
@@ -96,11 +96,5 @@
     <CustomBuild Include="map.c">
       <Filter>Source Files</Filter>
     </CustomBuild>
-    <CustomBuild Include="cgroup_sock_addr.c">
-      <Filter>Source Files</Filter>
-    </CustomBuild>
-    <CustomBuild Include="printk_legacy.c">
-      <Filter>Source Files</Filter>
-    </CustomBuild>
```
Comment (dthaler) [inline tests/sample/sample.vcxproj.filters:104]: Why are these removed?

**Review by dthaler — APPROVED** [review]

File: tests/sample/encap_reflect_packet.c (hunk 9 above)
Comment (saxena-anurag) [reply tests/sample/encap_reflect_packet.c:6]: Are we moving to LLVM v13 with this PR?

**Review by saxena-anurag — APPROVED** [review]
